pkg/query-service/app/traces/v4/query_builder.go

```diff
@@ -0,0 +1,415 @@
+		filterSubQuery = filterSubQuery + " AND " + emptyValuesInGroupByFilter
+	}
+
+	resourceSubQuery, err := resource.BuildResourceSubQuery("signoz_traces", "distributed_traces_v3_resource", bucketStart, bucketEnd, mq.Filters, mq.GroupBy, mq.AggregateAttribute, false)
+	if err != nil {
+		return "", err
+	}
+	// join both the filter clauses
+	if resourceSubQuery != "" {
+		filterSubQuery = filterSubQuery + " AND (resource_fingerprint GLOBAL IN " + resourceSubQuery + ")"
+	}
+
+	// timerange will be sent in epoch millisecond
+	selectLabels := getSelectLabels(mq.GroupBy)
+	if selectLabels != "" {
+		selectLabels = selectLabels + ","
+	}
+
+	orderBy := orderByAttributeKeyTags(panelType, mq.OrderBy, mq.GroupBy)
+	if orderBy != "" {
+		orderBy = " order by " + orderBy
+	}
+
+	if mq.AggregateOperator == v3.AggregateOperatorNoOp {
+		var query string
+		if panelType == v3.PanelTypeTrace {
+			withSubQuery := fmt.Sprintf(constants.TracesExplorerViewSQLSelectWithSubQuery, constants.SIGNOZ_TRACE_DBNAME, constants.SIGNOZ_SPAN_INDEX_V3_LOCAL_TABLENAME, timeFilter, filterSubQuery)
+			withSubQuery = tracesV3.AddLimitToQuery(withSubQuery, mq.Limit)
+			if mq.Offset != 0 {
+				withSubQuery = tracesV3.AddOffsetToQuery(withSubQuery, mq.Offset)
+			}
+			query = fmt.Sprintf(constants.TracesExplorerViewSQLSelectBeforeSubQuery, constants.SIGNOZ_TRACE_DBNAME, constants.SIGNOZ_SPAN_INDEX_V3) + withSubQuery + ") " + fmt.Sprintf(constants.TracesExplorerViewSQLSelectAfterSubQuery, constants.SIGNOZ_TRACE_DBNAME, constants.SIGNOZ_SPAN_INDEX_V3, timeFilter)
+		} else if panelType == v3.PanelTypeList {
+			if len(mq.SelectColumns) == 0 {
+				return "", fmt.Errorf("select columns cannot be empty for panelType %s", panelType)
+			}
+			// add it to the select labels
+			selectLabels = getSelectLabels(mq.SelectColumns)
+			queryNoOpTmpl := fmt.Sprintf("SELECT timestamp as timestamp_datetime, spanID, traceID,%s ", selectLabels) + "from " + constants.SIGNOZ_TRACE_DBNAME + "." + constants.SIGNOZ_SPAN_INDEX_V3 + " where %s %s" + "%s"
+			query = fmt.Sprintf(queryNoOpTmpl, timeFilter, filterSubQuery, orderBy)
+		} else {
+			return "", fmt.Errorf("unsupported aggregate operator %s for panelType %s", mq.AggregateOperator, panelType)
+		}
+		return query, nil
+		// ---- NOOP ends here ----
+	}
+
+	having := tracesV3.Having(mq.Having)
+	if having != "" {
+		having = " having " + having
+	}
+
+	groupBy := tracesV3.GroupByAttributeKeyTags(panelType, options.GraphLimitQtype, mq.GroupBy...)
+	if groupBy != "" {
+		groupBy = " group by " + groupBy
+	}
+
+	aggregationKey := ""
+	if mq.AggregateAttribute.Key != "" {
+		aggregationKey = getColumnName(mq.AggregateAttribute)
+	}
+
+	var queryTmpl string
+	if options.GraphLimitQtype == constants.FirstQueryGraphLimit {
+		queryTmpl = "SELECT"
+	} else if panelType == v3.PanelTypeTable {
+		queryTmpl =
+			"SELECT "
+		// step or aggregate interval is whole time period in case of table panel
+		step = (tracesEnd - tracesStart) / 1000000000
```

Comment:
What is this doing here? If it is used, it will give incorrect results (i.e two values for the group instead of one whole value) in certain cases. For example, take data ranging from 12:15 to 12:45, the toStartOfInterval will on point for `12:15-12:30` and another for `12:30-12:45`.

Reply:
Yeah it's not used, but it used to be present in the old QB

https://github.com/SigNoz/signoz/blob/969ac5028e71ea2445d8beef11d39278acb48df5/pkg/query-service/app/traces/v3/query_builder.go#L259 .

Let me remove it.